### ISSUE TITLE
[selectors-4] Issue 11: Introduce pseudo-class matching when user changed the value of an input

### ISSUE THREAD

**Issue by SebastianZ** [w3c/csswg-drafts#1533]
Back in 2014 it was [suggested to add a pseudo-class matching when a user changed the value of an element](https://lists.w3.org/Archives/Public/www-style/2014Feb/0511.html). This topic was also already [discussed earlier by the CSSWG](https://lists.w3.org/Archives/Public/www-style/2012Aug/0749.html).

I want to pick up that discussion again. The idea previously was to check whether the current value differs from the default value, expressed in JavaScript syntax, where `el.value !== el.defaultValue`, `el.checked !== el.defaultChecked`, currently selected options different from default selected options, etc., depending on the type of input.

Suggested names so far are `:changed`, `:edited`, `:dirty`, `:user-dirtied` and `:user-interacted`.

This would make `:user-invalid` redundant, as it could be expressed by `:changed:invalid`.

Sebastian

**Comment by fantasai** [comment]
@SebastianZ Could you give some specific use cases, other than those covered by `:user-invalid`? :)

**Comment by tomhodgins** [comment]
Oooh what a fun idea! Here's a quick demo of what using `:modified` in CSS stylesheets might feel like by using `el.value !== el.defaultValue` logic in JavaScript with a made-up `[--modified]` selector in CSS that we can grab onto to implement the functionality: https://codepen.io/tomhodgins/pen/YdKveq

```css
input[--modified] {
  background: lime;
}
```

It's kind of fun, and does fill a current gap with `[value=""]` and `<input>` elements between the `value` and the [dirty value](https://html.spec.whatwg.org/#concept-fe-dirty).

**Comment by zoliszabo** [comment]
Given the above linked [dirty value](https://html.spec.whatwg.org/#concept-fe-dirty), IMHO the best name for this pseudo-class would be **`:dirty`**.

The [HTML Live Standard](https://html.spec.whatwg.org/) defines the dirty flag as
> `input` and `textarea` elements have a dirty value flag. This is used to track the interaction between the value and default value. If it is false, value mirrors the default value. If it is true, the default value is ignored.

while `select` elements have the [concept of dirtiness](https://html.spec.whatwg.org/#concept-option-dirtiness) on their `option` descendants:
> The dirtiness of an `option` element is a boolean state, initially false. It controls whether adding or removing the `selected` content attribute has any effect.

In this case dirtiness should bubble up from the `option` elements to their parent `select` element.

**Comment by htstudios** [comment]
👍 for :dirty (aka different from what has been submitted by the http response)

**Comment by htstudios** [comment]
... and :modified would NOT be the same.

Modified could mean the same as unix "touched" which could mean the value would be the same but you touched it.

**Comment by SebastianZ** [comment]
As there was a lot of positive feedback and possible use cases mentioned, I set this as Agenda+.

Sebastian

**Comment by nt1m** [comment]
With `:placeholder-shown`, `:checked`, `:indeterminate`, etc. I'm not sure what's the appeal for this. But I think the bigger issue I have with this, is what the "default" state is.

**Comment by zoliszabo** [comment]
> With `:placeholder-shown`, `:checked`, `:indeterminate`, etc. I'm not sure what's the appeal for this. But I think the bigger issue I have with this, is what the "default" state is.

There are cases when you want to mark/highlight the input fields that were modified since page load. While this is easy to do for checkboxes/radio buttons (e.g. `input[checked]:not(:checked) {...}` and `input:not([checked]):checked {...}`), there isn't such a "straightforward" solution in the case of text inputs, textareas or selects.

In the context of `:dirty`, "default" would mean the value the input field had when the page was loaded (it might be empty or not). It is a good point, however, that in the case of SPAs there should be a way to "mark" a new value as default.

**Comment by zoliszabo** [comment]
> Wouldn't this be equivalent to `:is(:user-valid, :user-invalid)` or somesuch?

Would you like to shortly explain how this would work?

**Comment by emilio** [comment]
https://drafts.csswg.org/selectors-4/#user-pseudos says:

> The :user-invalid and the :user-valid pseudo-classes represent an element with incorrect or correct input, respectively, but only after the user has significantly interacted with it.

So `:is(:user-valid, :user-invalid)` should cover all elements (both correct and incorrect) that the user has interacted with. So it's not exactly a `!= default`, but it seems to address the relevant use-cases which are more on the line of "things the user has interacted with" rather than the specific "it's not the default value".

**Comment by zoliszabo** [comment]
Thanks! On a quick read I failed to see that `:is(:user-valid, :user-invalid)` was an actual selector and not an enumeration of the `:is(:user-valid)` and `:is(:user-invalid)` selectors.

You are right that `:is(:user-valid, :user-invalid)` covers all "significantly interacted with" elements, but there is at least one drawback which makes it incompatible with `:dirty`:

https://drafts.csswg.org/selectors-4/#user-pseudos also says:
> For example, the input in the following document fragment would match '':invalid'' as soon as the page is loaded (because it the initial value violates the max-constraint), but it won't match '':user-invalid'' until the user significantly interacts with the element, or attempts to submit the form it's part of.
```HTML
<form>
    <label>
        Volume:
        <input name='vol' type=number min=0 max=10 value=11>
    </label>
    ...
</form>
```
Basically the `:is(:user-valid, :user-invalid)` would kick in on form submit even if there was no interaction with the elements.

This being said, there probably is a combination of `:user-valid`, `:user-invalid`, `:valid`, `:invalid` that would do the job for each of the most common real-life use-cases, but I would still prefer the more straightforward `:dirty` 😀

**Comment by jensimmons** [comment]
I'm _really_ not a fan of the name `:dirty`. It seems very inconsistent with the kinds of names in CSS. It feels like it carries a politically problematic judgement. And it does not convey the meaning of the term — it’s most important job

**Comment by scottaohara** [comment]
seems like this would be useful information to expose beyond just through styles.  i'd be concerned if a page exposed a bunch of visually changed fields that indicated "these fields have changed", but there is no programmatic exposure of such for people who can't perceive the style changes.   Existing pseudo-class styling like `:placeholder-shown`, `:invalid`, `:user-valid`, `:empty`, etc., seem like they cover a lot of use cases and are associated with a programmatically exposed state - but just "this was changed from one valid value to another", how might that be relayed?

**Comment by jpzwarte** [comment]
So wanted to add my 2 cents: I just ran into a scenario where I have a FACE web component (Form Associated Custom Element) that is `required`. I want to change the styling of the component when `:invalid`, *but only* after the user has interacted with the component.

```css
:host(:invalid:user-interacted) { 
  border-color: red; 
}
```

Btw, I am more familiar with the term `:touched` instead of `:user-interacted` (Angular & angular.js use the term "touched" for example). But perhaps touched can be confusing: it implies a user has touched (using a finger?) the component.

**Comment by SebastianZ** [comment]
> Btw, I am more familiar with the term `:touched` instead of `:user-interacted` (Angular & angular.js use the term "touched" for example). But perhaps touched can be confusing: it implies a user has touched (using a finger?) the component.

There are probably many people out there that understand the concept of `:touched`, though I'd still rule it out for the reason you provided.

Sebastian

**Comment by jensimmons** [comment]
> :invalid:user-interacted

Isn't this usecase already covered by :user-invalid? And it's twin, :user-valid.

**Comment by jpzwarte** [comment]
@jensimmons Yes, the more I read about it, the more I think `:user-invalid` is better suited. One question I still have though: what about if you *haven't* interacted with the form element, but you've called `form.requestSubmit()`. How would you style the invalid form element after that? Does `:user-invalid` still apply in that case? (`:user-interacted` wouldn't either afaict).

**Comment by jacobrask** [comment]
What some in this issue seem to argue for is rather the :blank selector. To create the Material style inputs with labels that are first positioned inside the input but then animate away, what you’re interested in is if the input is blank or not. If the default value is non-empty you also want to move the label.

[Developers](https://github.com/finnhvman/matter) currently need to use :placeholder-shown in combination with a placeholder attribute with whitespace , which is not intuitive and is one of the few remaining reasons why we need JS for our inputs

**Comment by scottaohara** [comment]
>Does :user-invalid still apply in that case? 

if a form attempts to be submitted but a user hasn't filled out required fields, then i don't see why that wouldn't be considered user-invalid.  They did not meet all requirements for form submission.

**Comment by waterplea** [comment]
We need both `:dirty` and `:touched`. Currently `:user-invalid` does not trigger when I focus a required field and blur it. This is suboptimal when just tabbing through a form — user should be notified in this case that they missed out on a required value. Combination of those selectors would make it possible.

**Comment by bramus** [comment]
Angular [defines](https://v17.angular.io/api/forms/AbstractControl#properties) the following:

State | Description
-- | --
`pristine` | A control is pristine if the user has not yet changed the value in the UI.
`dirty` | A control is dirty if the user has changed the value in the UI.
`touched` | True if the control is marked as touched. A control is marked touched once the user has triggered a blur event on it.
`untouched` | True if the control has not been marked as touched. A control is untouched if the user has not yet triggered a blur event on it.

These terms also appear in popular libraries such as React Hook Form, see [its `useFormState` hook](https://react-hook-form.com/docs/useformstate).

These all seems like reasonable pseudos to add to CSS.

Regarding the relation to `user-valid`/`user-invalid`:
- The `pristine`/`dirty` pair triggers much earlier
- The `touched`/`untouched` pair triggers much later

Also note that an input can be both `touched` and `pristine` in case the user has edited the value, but then changed it back to its original value.

**Comment by emilio** [comment]
If user-invalid has suboptimal behavior we should fix that, imo

**Comment by waterplea** [comment]
> If user-invalid has suboptimal behavior we should fix that, imo

I'm trying here:
https://issues.chromium.org/issues/365802576
https://bugs.webkit.org/show_bug.cgi?id=279503

Haven't filed one for Firefox

Having native CSS selectors for controls states would allow better flexibility though, as you can see Chromium guy who implemented it argues that current behavior is correct and in accordance with the specs. Current behavior is unacceptable per our design guidelines, which is the same as how Angular does it — show invalid controls after user visited them, not edited them.

**Comment by emilio** [comment]
> Having native CSS selectors for controls states would allow better flexibility though, as you can see Chromium guy who implemented it argues that current behavior is correct and in accordance with the specs.

That is true. But the html spec could be changed with agreement from vendors and tests.

> Current behavior is unacceptable per our design guidelines, which is the same as how Angular does it — show invalid controls after user visited them, not edited them.

That's how Firefox used to do it iirc, before I changed it to follow the current spec.

**Comment by waterplea** [comment]
The current spec only says 3 things:

1. It should match "only after the user has significantly interacted with it" without specifying criteria of significance.
2. It should match after form submit attempt
3. And "user-agents may allow them to match such elements at other times, as would be appropriate for highlighting an error to the user"

I believe focusing and leaving an empty required input is exactly "appropriate for highlighting an error to the user" 🙂 I would even go as far as to say focusing and leaving any invalid input should highlight it for user, but I can see how this could be up for a debate so it would be great if we could personalize this to our liking with new selectors.

**Comment by svgeesus** [comment]
So, does the Selectors 4 spec need changes as a result of this discussion? I see people saying that the current text is correct.

**Comment by svgeesus** [comment]
@SebastianZ since you raised the issue

**Comment by oleastre** [comment]
I have a use case for a `:dirty` like selector: change buttons appearance based on form state.
In my use case, I have a wizard like behaviour where a user has to go through several pages, validate or modify input and got to the next page.
Current implementation uses javascript to catch "change" events, but I could imagine having something like:
```html
<form>
  <input name="test" value="Current">
  <button type="submit">Save</button>
  <a href="next">Next</a>
</form>
```

```css
form {
  button { display: none }
  a {display: block}
}
form:has(input:dirty) {
  button { display: block }
  a {display: none}
}
```
=> when the user edits a value, the "Save" button appears otherwise, the user gets a link to the next page.
